example/reduce/src/kernel.hpp

```diff
@@ -0,0 +1,170 @@
+//!
+//! \tparam TBlockSize The block size.
+//! \tparam T The data type.
+//! \tparam TFunc The Functor type for the reduction function.
+template <uint32_t TBlockSize, typename T, typename TFunc>
+struct ReduceKernel
+{
+    ALPAKA_NO_HOST_ACC_WARNING
+
+    //-----------------------------------------------------------------------------
+    //! The kernel entry point.
+    //!
+    //! \tparam TAcc The accelerator environment.
+    //! \tparam TElem The element type.
+    //! \tparam TIdx The index type.
+    //!
+    //! \param acc The accelerator object.
+    //! \param source The source memory.
+    //! \param destination The destination memory.
+    //! \param n The problem size.
+    //! \param func The reduction function.
+    template <typename TAcc, typename TElem, typename TIdx>
+    ALPAKA_FN_ACC auto operator()(TAcc const &acc,
+                                  TElem const *const source,
+                                  TElem *destination,
+                                  TIdx const &n,
+                                  TFunc func) const -> void
+    {
+        auto &sdata(
+            alpaka::block::shared::st::allocVar<cheapArray<T, TBlockSize>,
+                                                __COUNTER__>(acc));
+
+        const uint32_t blockIndex(
+            alpaka::idx::getIdx<alpaka::Grid, alpaka::Blocks>(acc)[0]);
+        const uint32_t threadIndex(
+            alpaka::idx::getIdx<alpaka::Block, alpaka::Threads>(acc)[0]);
+        const uint32_t gridDimension(
+            alpaka::workdiv::getWorkDiv<alpaka::Grid, alpaka::Blocks>(acc)[0]);
+
+        // equivalent to blockIndex * TBlockSize + threadIndex
+        const uint32_t linearizedIndex(
+            alpaka::idx::getIdx<alpaka::Grid, alpaka::Threads>(acc)[0]);
+
+        typename GetIterator<T, TElem, TAcc>::Iterator it(
+            acc, source, linearizedIndex, gridDimension * TBlockSize, n);
+
+        T result;
+
+        if (threadIndex < n)
+            result = *(it++); // avoids using the
+                              // neutral element of specific
+
+        // --------
+        // Level 1: grid reduce, reading from global memory
+        // --------
+
+        // reduce per thread with increased ILP by 4x unrolling sum.
+        // the thread of our block reduces its 4 grid-neighbored threads and
+        // advances by grid-striding loop (maybe 128bit load improve perf)
+
+        while (it + 3 < it.end())
+        {
+            result = func(
+                func(func(result, func(*it, *(it + 1))), *(it + 2)), *(it + 3));
+            it += 4;
+        }
+
+        // doing the remaining blocks
+        while (it < it.end())
+            result = func(result, *(it++));
+
+        if (threadIndex < n)
+            sdata[threadIndex] = result;
+
+        alpaka::block::sync::syncBlockThreads(acc);
+
+        // --------
+        // Level 2: block + warp reduce, reading from shared memory
+        // --------
+
+#pragma unroll
```

Comment:
`#pragma unroll` is not supported by all compilers (as the CI shows). Try to use `ALPAKA_UNROLL` from `alpaka/core/Unroll.hpp` instead. It might be a bit out of date but it hopefully works.